### ISSUE TITLE
Question: How to write csv files with comma as decimal separator

### ISSUE THREAD

**Issue by deveborn** [jtablesaw/tablesaw#845]
Is there a simple way to write csv files with comma as decimal separator? I have not find a way to do that through CsvWriteOptions? Most European countries use comma as decimal separator and to be able to open files in Excel (without extra fiddeling) you want to use comma as separator in your outputs. 

Thanks

David

**Comment by lwhite1** [comment]
This is not currently possible. It would minimally require adding a locale option to CsvWriteOptions

**Comment by deveborn** [comment]
I would be glad to contribute on this if I get pointed in right
direction.  I took a closer look to this yesterday. My plan was to create
an option to add NumberFormatters. One for decimal numbers (double and
float columns) and one for whole numbers (shortColums LongColumns and
IntColumns). I would also add a shortcut to request a particular decimal
delimiter (which is then adding a NumberFormatter for decimal numbers with
the decided decimial delimiter). Do you think that this would be a
reasonable solution?

Den tors 21 jan. 2021 kl 01:17 skrev Larry White <notifications@github.com>:

**Comment by lwhite1** [comment]
That sounds pretty good. I would be glad to have a solution for this.

I think it might be best to add number formatters as a map, maybe Map<String, NumberFormatter> where the string is a column name. 

It's hard to imagine a scenario where Date formats should be different from column to column, but numbers could easily be different, especially with tablesaw's aggregation and number map functions turning everything into Double Column. It's a pretty common issue for me to sum an int column and then have the sum written as a decimal number that I need to fix.  

I think this would also solve an open issue for supporting currency formats. Again, you might have currencies and other numbers in the same table. 

I'm not completely satisfied with the shortcut solution because there might be other changes you want to apply to everything, like using a . for a hundreds separator, or changing the rounding mode. I also think if the 'shortcut' is built into the same method as the main solution it will be less confusing to the user because all to documentation is in one place.. 

Maybe we could have a standard where a map entry with a key of empty string is applied to any Number Column that isn't explicitly named? IDK what happens if you try to apply a format with a decimal delimiter to an integer, though. 

 What do you think?

**Comment by benmccann** [comment]
Maybe we could take a function as an option like `String format(ColumnType columnType; String columnName; int row; int column; Object value)` and then it can be customized in almost any way by the user

**Comment by lwhite1** [comment]
@benmccann I have mixed feelings. That solves the complete problem, but seems a little heavyweight to me. The main advantage, I think, is that it covers booleans, which aside from numbers and time is the main thing that people write in multiple ways. Strings basically are their string format, so you're really changing the data if you reformat strings. Seems like the kind of thing to be done as a map function on the column. 

The disadvantage is that you're writing more code to handle the common cases. As mentioned above, I think all the dates in a file should be formatted the same way, so that's done with a single option for the file. WIth number formatting, you can just name the column(s) to be formatted without dealing with column types or wrapping the formatter in a function. 

I'd probably have to see example code side-by-side to really understand how big the difference is.

**Comment by deveborn** [comment]
@lwhite1 One formatter per column gives better control of course and as you state there are use cases. I agree with you that the shortcut can be a bit confusing. The reason I wanted it was the somewhat tedious way of creating a number formatter with custom decimal separator (but it is not a big deal):

```      
        DecimalFormatSymbols symbols = new DecimalFormatSymbols(Locale.getDefault());
        symbols.setDecimalSeparator(',');
        DecimalFormat decimalFormat=new DecimalFormat();
        decimalFormat.setDecimalFormatSymbols(symbols);
        decimalFormat.setGroupingUsed(false);
```
May I suggest that, instead of having a map with special treatment of empty string, one could have two arguments in the setter. One for a default numberformatter and one map as you suggested with column specific formatters. Like so...
``` 
public CsvWriteOptions.Builder numberFormatter(NumberColumnFormatter defaultFormatter, Map<String,NumberColumnFormatter> columnSpecificFormatters)
```
Ben, do you suggest a custom function as a complement to column formatters or as an exchange?

**Comment by lwhite1** [comment]
@deveborn I guess I assumed you would be changing the locale to one where "," was the default decimal indicator and you would be done. 

e.g. 

```
    NumberFormat df = NumberFormat.getInstance(Locale.GERMANY);
    df.format(123.34);  // prints "123,34" on my machine. 
 
```

**Comment by lwhite1** [comment]
Also, @deveborn I think NumberColumnFormatter was a case of me overthinking things to save a few keystrokes. It would probably be best to build the API on NumberFormat directly.

**Comment by benmccann** [comment]
Yeah, I think you're right my original idea was a bit heavyweight. 

Really you could convert any column to a string column and do an arbitrary formatting. So I guess the question is solely about whether that's too verbose and an easier API would be helpful. Setting a `Locale` and using that for all number and date formatting by default could be an option as well

**Comment by deveborn** [comment]
@lwhite1 That's true. Probably changing Locale and setting grouping to false will be a common case. I thought that ColumnNumberFormat maybe were useful for proper handling of missing values? But for example one can use NumberFormat directly and fall back on "standard writing" when NumberFormat throws exception.

@benmccann If a Locale should be used per default I guess that it may have some impacts on code written in earlier versions (if I am right, there is no Formatter involved at all for the NumberColumns at the moment). I do not know if compatibility in that sense is important or not. Setting a Locale however, do not it self serve the per column precision Larry was asking for earlier. But I guess it can be combined with an option for setting column specific formatters.

**Comment by lwhite1** [comment]
@deveborn Oh yeah. missing values. I guess we should consider it as having people write their own missing value handling isn't great. 

Locale is a step in the right direction, but by itself it won't do formatting for currency or things that are not locale related like whether to use separators for hundreds, or to format a double column as integers. 

I kinda feel like using a map is good, but maybe also provide a default for integral cols and another for floating point? If they're null then no formatting is applied unless the column is mentioned in the map?

**Comment by lwhite1** [comment]
@deveborn Any thoughts on when you might have something? No rush, but I'm thinking there's enough changes here to release if it will be a while.

**Comment by deveborn** [comment]
Thanks for asking. I am working on it but you do not have to wait for me.
Eventually I manage to pull something tomorrow but if I don't, it may be
weeks away because of work and family. Moreover, since it actually will be
my first pull request ever, you should not have to high expectations! It is
a learning experience for me but I still hope you will get something out of
it!

Den lör 30 jan. 2021 kl 19:12 skrev Larry White <notifications@github.com>:

**Comment by lwhite1** [comment]
That sounds fine. Thanks for your help.

**Comment by lwhite1** [comment]
I renamed this to remove the reference to reading data as it was only mentioned in the title.  The ability to write data formatted in this way is implemented in #914.
If reading data from CSV with ',' as the decimal separator is still a concern, please open a separate issue for that. Thanks.
